compiler/circle-opselector/src/OpSelector.cpp

```diff
@@ -0,0 +1,505 @@
+  // node's inputs right after creating the node.
+  for (uint32_t i = 0; i < operators.size(); ++i)
+  {
+    const circle::OperatorT &op = *operators[i];
+    circle::BuiltinOperator builtincode = reader.builtin_code(op);
+
+    if (const auto *builder = source.lookup(builtincode))
+    {
+      luci::GraphBuilder::ValidateArgs args(op, reader);
+      if (!builder->validate(args))
+      {
+        throw oops::UserExn("Invalid operator", reader.opcode_name(op));
+      }
+
+      auto built_op = builder->build(op, &gb_context);
+      set_node_id(built_op, i);
+
+      add_origin(built_op, luci::single_origin(i, built_op->name()));
+    }
+    else
+    {
+      throw oops::UserExn("Not supported", reader.opcode_name(op));
+    }
+  }
+
+  // graph outputs
+  for (auto output : reader.outputs())
+  {
+    const circle::TensorT &tensor = *tensors[output];
+
+    auto output_node = graph->nodes()->create<luci::CircleOutput>();
+    assert(output_node != nullptr);
+    auto output_from = nodefinder->node(output);
+    if (output_from != nullptr)
+      output_node->from(output_from);
+    else
+    {
+      // NOTE loco::Graph requires all input node(s) to a node should exist.
+      //      Here, CircleOutput needs an input node.
+      //      We add a dummy node to make it happy.
+      auto output_dummy = graph->nodes()->create<luci::CircleOutputDummy>();
+      assert(output_dummy != nullptr);
+      output_node->from(output_dummy);
+
+      luci::copy_tensor_attributes(tensor, output_dummy);
+      if (tensors_ptr->Get(output)->shape() == nullptr)
+        output_dummy->shape_status(luci::ShapeStatus::NOSHAPE);
+      else
+        output_dummy->shape_status(luci::ShapeStatus::VALID);
+    }
+
+    // set the graph output name and node object
+    auto graph_output = graph->outputs()->create();
+    std::string tname = luci::tensor_name(tensor);
+    assert(tname.length() > 0);
+    graph_output->name(tname);
+
+    luci::copy_tensor_attributes(tensor, output_node);
+
+    // Set GraphInputOutputIndex for graph
+    output_node->index(graph_output->index());
+
+    assert(tensor.shape_signature.size() == 0 ||
+           tensor.shape_signature.size() == tensor.shape.size());
+
+    // Shape of Output
+    auto output_shape = std::make_unique<loco::TensorShape>();
+    const std::vector<int32_t> &output_dims = tensor.shape; // in NHWC
+    output_shape->rank(output_dims.size());
+    for (uint32_t r = 0; r < output_dims.size(); ++r)
+    {
+      if (tensor.shape_signature.size() > 0 && tensor.shape_signature.at(r) == -1)
+        output_shape->dim(r).unset();
+      else
+        output_shape->dim(r).set(output_dims[r]);
+    }
+    graph_output->shape(std::move(output_shape));
+
+    // Data type
+    auto dtype = luci::luci_datatype(tensor.type);
+    graph_output->dtype(dtype);
+  }
+}
+
+std::unique_ptr<luci::Module>
+OpSelector::select_nodes(std::vector<const luci::CircleNode *> selected_nodes)
```

Comment:
this method seems very long. it would be better to split to small functions by its functionality.